File: ios/Sasquatsh/Models/Enums.swift
import Foundation

enum DifficultyLevel: String, Codable, CaseIterable, Identifiable {
    case beginner
    case intermediate
    case advanced

    var id: String { rawValue }

    var displayName: String {
        rawValue.capitalized
    }
}

enum EventStatus: String, Codable, CaseIterable, Identifiable {
    case draft
    case published
    case cancelled
    case completed

    var id: String { rawValue }

    var displayName: String {
        rawValue.capitalized
    }
}

enum ItemCategory: String, Codable, CaseIterable, Identifiable {
    case food
    case drinks
    case supplies
    case other

    var id: String { rawValue }

    var displayName: String {
        rawValue.capitalized
    }
}

enum GameCategory: String, Codable, CaseIterable, Identifiable {
    case strategy
    case party
    case cooperative
    case deckbuilding
    case workerplacement
    case areacontrol
    case dice
    case trivia
    case roleplaying
    case miniatures
    case card
    case family
    case abstract
    case other

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .deckbuilding: return "Deck Building"
        case .workerplacement: return "Worker Placement"
        case .areacontrol: return "Area Control"
        case .roleplaying: return "Roleplaying"
        default: return rawValue.capitalized
        }
    }
}

enum GroupType: String, Codable, CaseIterable, Identifiable {
    case geographic
    case interest
    case both

    var id: String { rawValue }

    var displayName: String {
        rawValue.capitalized
    }
}

enum MemberRole: String, Codable, CaseIterable, Identifiable {
    case owner
    case admin
    case member

    var id: String { rawValue }

    var displayName: String {
        rawValue.capitalized
    }
}

enum JoinPolicy: String, Codable, CaseIterable, Identifiable {
    case open
    case request
    case invite_only

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .open: return "Open"
        case .request: return "Request to Join"
        case .invite_only: return "Invite Only"
        }
    }
}

enum PlanningStatus: String, Codable, CaseIterable, Identifiable {
    case open
    case finalized
    case cancelled

    var id: String { rawValue }

    var displayName: String {
        rawValue.capitalized
    }
}

enum JoinRequestStatus: String, Codable {
    case pending
    case approved
    case rejected
}

enum SubscriptionTier: String, Codable, CaseIterable, Identifiable {
    case free, basic, pro, premium

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .free: return "Free"
        case .basic: return "Basic"
        case .pro: return "Pro"
        case .premium: return "Premium"
        }
    }

    var priceLabel: String {
        switch self {
        case .free: return "$0"
        case .basic: return "$7.99/mo"
        case .pro: return "$14.99/mo"
        case .premium: return "Custom"
        }
    }

    var rank: Int {
        switch self {
        case .free: return 0
        case .basic: return 1
        case .pro: return 2
        case .premium: return 3
        }
    }
}

enum SubscriptionStatus: String, Codable {
    case active
    case pastDue = "past_due"
    case canceled
    case incomplete
}

enum InvoiceStatus: String, Codable {
    case paid, open, draft, void, uncollectible
}

enum EventLocationStatus: String, Codable {
    case pending
    case approved
    case rejected
}

enum GameSystem: String, Codable, CaseIterable, Identifiable {
    case boardGame = "board_game"
    case mtg
    case pokemonTcg = "pokemon_tcg"
    case yugioh
    case warhammer40k

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .boardGame: return "Board Game"
        case .mtg: return "Magic: The Gathering"
        case .pokemonTcg: return "Pokémon TCG"
        case .yugioh: return "Yu-Gi-Oh!"
        case .warhammer40k: return "Warhammer 40K"
        }
    }

    var shortName: String {
        switch self {
        case .boardGame: return "Board Game"
        case .mtg: return "MTG"
        case .pokemonTcg: return "Pokémon"
        case .yugioh: return "Yu-Gi-Oh!"
        case .warhammer40k: return "40K"
        }
    }

    var iconName: String {
        switch self {
        case .boardGame: return "dice"
        case .mtg: return "rectangle.portrait.on.rectangle.portrait.fill"
        case .pokemonTcg: return "bolt.circle.fill"
        case .yugioh: return "star.circle.fill"
        case .warhammer40k: return "shield.fill"
        }
    }

    var logoAssetName: String {
        switch self {
        case .boardGame: return "Logo"
        case .mtg: return "mtg-logo"
        case .pokemonTcg: return "pokemon-logo"
        case .yugioh: return "yugioh-logo"
        case .warhammer40k: return "warhammer40k-logo"
        }
    }
}

enum AppTimezone: String, CaseIterable, Identifiable {
    case eastern = "America/New_York"
    case central = "America/Chicago"
    case mountain = "America/Denver"
    case arizona = "America/Phoenix"
    case pacific = "America/Los_Angeles"
    case alaska = "America/Anchorage"
    case hawaii = "Pacific/Honolulu"
    case uk = "Europe/London"
    case centralEurope = "Europe/Paris"
    case japan = "Asia/Tokyo"
    case sydney = "Australia/Sydney"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .eastern: return "Eastern (ET)"
        case .central: return "Central (CT)"
        case .mountain: return "Mountain (MT)"
        case .arizona: return "Arizona (MST)"
        case .pacific: return "Pacific (PT)"
        case .alaska: return "Alaska (AKT)"
        case .hawaii: return "Hawaii (HST)"
        case .uk: return "UK (GMT/BST)"
        case .centralEurope: return "Central Europe (CET)"
        case .japan: return "Japan (JST)"
        case .sydney: return "Sydney (AEST)"
        }
    }
}

File: android/app/src/main/java/com/sasquatsh/app/models/Enums.kt
package com.sasquatsh.app.models

import com.squareup.moshi.Json

enum class GameSystem(val value: String) {
    @Json(name = "board_game") BOARD_GAME("board_game"),
    @Json(name = "mtg") MTG("mtg"),
    @Json(name = "pokemon_tcg") POKEMON_TCG("pokemon_tcg"),
    @Json(name = "yugioh") YUGIOH("yugioh"),
    @Json(name = "warhammer40k") WARHAMMER_40K("warhammer40k");

    val displayName: String
        get() = when (this) {
            BOARD_GAME -> "Board Game"
            MTG -> "Magic: The Gathering"
            POKEMON_TCG -> "Pokémon TCG"
            YUGIOH -> "Yu-Gi-Oh!"
            WARHAMMER_40K -> "Warhammer 40K"
        }

    val shortName: String
        get() = when (this) {
            BOARD_GAME -> "Board Game"
            MTG -> "MTG"
            POKEMON_TCG -> "Pokémon"
            YUGIOH -> "Yu-Gi-Oh!"
            WARHAMMER_40K -> "40K"
        }

    val iconResName: String
        get() = when (this) {
            BOARD_GAME -> "ic_dice"
            MTG -> "ic_cards"
            POKEMON_TCG -> "ic_bolt_circle"
            YUGIOH -> "ic_star_circle"
            WARHAMMER_40K -> "ic_shield"
        }

    val logoAssetName: String
        get() = when (this) {
            BOARD_GAME -> "Logo"
            MTG -> "mtg-logo"
            POKEMON_TCG -> "pokemon-logo"
            YUGIOH -> "yugioh-logo"
            WARHAMMER_40K -> "warhammer40k-logo"
        }

    companion object {
        fun fromValue(value: String): GameSystem? =
            entries.find { it.value == value }
    }
}

enum class SubscriptionTier(val value: String) {
    @Json(name = "free") FREE("free"),
    @Json(name = "basic") BASIC("basic"),
    @Json(name = "pro") PRO("pro"),
    @Json(name = "premium") PREMIUM("premium");

    val displayName: String
        get() = when (this) {
            FREE -> "Free"
            BASIC -> "Basic"
            PRO -> "Pro"
            PREMIUM -> "Premium"
        }

    val priceLabel: String
        get() = when (this) {
            FREE -> "$0"
            BASIC -> "$4.99/mo"
            PRO -> "$7.99/mo"
            PREMIUM -> "Custom"
        }

    val rank: Int
        get() = when (this) {
            FREE -> 0
            BASIC -> 1
            PRO -> 2
            PREMIUM -> 3
        }

    companion object {
        fun fromValue(value: String): SubscriptionTier? =
            entries.find { it.value == value }
    }
}

enum class GameCategory(val value: String) {
    @Json(name = "strategy") STRATEGY("strategy"),
    @Json(name = "party") PARTY("party"),
    @Json(name = "cooperative") COOPERATIVE("cooperative"),
    @Json(name = "deckbuilding") DECKBUILDING("deckbuilding"),
    @Json(name = "workerplacement") WORKERPLACEMENT("workerplacement"),
    @Json(name = "areacontrol") AREACONTROL("areacontrol"),
    @Json(name = "dice") DICE("dice"),
    @Json(name = "trivia") TRIVIA("trivia"),
    @Json(name = "roleplaying") ROLEPLAYING("roleplaying"),
    @Json(name = "miniatures") MINIATURES("miniatures"),
    @Json(name = "card") CARD("card"),
    @Json(name = "family") FAMILY("family"),
    @Json(name = "abstract") ABSTRACT("abstract"),
    @Json(name = "other") OTHER("other");

    val displayName: String
        get() = when (this) {
            DECKBUILDING -> "Deck Building"
            WORKERPLACEMENT -> "Worker Placement"
            AREACONTROL -> "Area Control"
            ROLEPLAYING -> "Roleplaying"
            else -> value.replaceFirstChar { it.uppercase() }
        }

    companion object {
        fun fromValue(value: String): GameCategory? =
            entries.find { it.value == value }
    }
}

enum class DifficultyLevel(val value: String) {
    @Json(name = "beginner") BEGINNER("beginner"),
    @Json(name = "intermediate") INTERMEDIATE("intermediate"),
    @Json(name = "advanced") ADVANCED("advanced");

    val displayName: String
        get() = value.replaceFirstChar { it.uppercase() }

    companion object {
        fun fromValue(value: String): DifficultyLevel? =
            entries.find { it.value == value }
    }
}

enum class EventStatus(val value: String) {
    @Json(name = "draft") DRAFT("draft"),
    @Json(name = "published") PUBLISHED("published"),
    @Json(name = "cancelled") CANCELLED("cancelled"),
    @Json(name = "completed") COMPLETED("completed");

    val displayName: String
        get() = value.replaceFirstChar { it.uppercase() }

    companion object {
        fun fromValue(value: String): EventStatus? =
            entries.find { it.value == value }
    }
}

enum class ItemCategory(val value: String) {
    @Json(name = "food") FOOD("food"),
    @Json(name = "drinks") DRINKS("drinks"),
    @Json(name = "supplies") SUPPLIES("supplies"),
    @Json(name = "other") OTHER("other");

    val displayName: String
        get() = value.replaceFirstChar { it.uppercase() }

    companion object {
        fun fromValue(value: String): ItemCategory? =
            entries.find { it.value == value }
    }
}

enum class GroupType(val value: String) {
    @Json(name = "geographic") GEOGRAPHIC("geographic"),
    @Json(name = "interest") INTEREST("interest"),
    @Json(name = "both") BOTH("both");

    val displayName: String
        get() = value.replaceFirstChar { it.uppercase() }

    companion object {
        fun fromValue(value: String): GroupType? =
            entries.find { it.value == value }
    }
}

enum class MemberRole(val value: String) {
    @Json(name = "owner") OWNER("owner"),
    @Json(name = "admin") ADMIN("admin"),
    @Json(name = "member") MEMBER("member");

    val displayName: String
        get() = value.replaceFirstChar { it.uppercase() }

    companion object {
        fun fromValue(value: String): MemberRole? =
            entries.find { it.value == value }
    }
}

enum class JoinPolicy(val value: String) {
    @Json(name = "open") OPEN("open"),
    @Json(name = "request") REQUEST("request"),
    @Json(name = "invite_only") INVITE_ONLY("invite_only");

    val displayName: String
        get() = when (this) {
            OPEN -> "Open"
            REQUEST -> "Request to Join"
            INVITE_ONLY -> "Invite Only"
        }

    companion object {
        fun fromValue(value: String): JoinPolicy? =
            entries.find { it.value == value }
    }
}

enum class PlanningStatus(val value: String) {
    @Json(name = "open") OPEN("open"),
    @Json(name = "finalized") FINALIZED("finalized"),
    @Json(name = "cancelled") CANCELLED("cancelled");

    val displayName: String
        get() = value.replaceFirstChar { it.uppercase() }

    companion object {
        fun fromValue(value: String): PlanningStatus? =
            entries.find { it.value == value }
    }
}

enum class JoinRequestStatus(val value: String) {
    @Json(name = "pending") PENDING("pending"),
    @Json(name = "approved") APPROVED("approved"),
    @Json(name = "rejected") REJECTED("rejected");

    companion object {
        fun fromValue(value: String): JoinRequestStatus? =
            entries.find { it.value == value }
    }
}

enum class SubscriptionStatus(val value: String) {
    @Json(name = "active") ACTIVE("active"),
    @Json(name = "past_due") PAST_DUE("past_due"),
    @Json(name = "canceled") CANCELED("canceled"),
    @Json(name = "incomplete") INCOMPLETE("incomplete");

    companion object {
        fun fromValue(value: String): SubscriptionStatus? =
            entries.find { it.value == value }
    }
}

enum class InvoiceStatus(val value: String) {
    @Json(name = "paid") PAID("paid"),
    @Json(name = "open") OPEN("open"),
    @Json(name = "draft") DRAFT("draft"),
    @Json(name = "void") VOID("void"),
    @Json(name = "uncollectible") UNCOLLECTIBLE("uncollectible");

    companion object {
        fun fromValue(value: String): InvoiceStatus? =
            entries.find { it.value == value }
    }
}

enum class EventLocationStatus(val value: String) {
    @Json(name = "pending") PENDING("pending"),
    @Json(name = "approved") APPROVED("approved"),
    @Json(name = "rejected") REJECTED("rejected");

    companion object {
        fun fromValue(value: String): EventLocationStatus? =
            entries.find { it.value == value }
    }
}

enum class AppTimezone(val value: String) {
    EASTERN("America/New_York"),
    CENTRAL("America/Chicago"),
    MOUNTAIN("America/Denver"),
    ARIZONA("America/Phoenix"),
    PACIFIC("America/Los_Angeles"),
    ALASKA("America/Anchorage"),
    HAWAII("Pacific/Honolulu"),
    UK("Europe/London"),
    CENTRAL_EUROPE("Europe/Paris"),
    JAPAN("Asia/Tokyo"),
    SYDNEY("Australia/Sydney");

    val displayName: String
        get() = when (this) {
            EASTERN -> "Eastern (ET)"
            CENTRAL -> "Central (CT)"
            MOUNTAIN -> "Mountain (MT)"
            ARIZONA -> "Arizona (MST)"
            PACIFIC -> "Pacific (PT)"
            ALASKA -> "Alaska (AKT)"
            HAWAII -> "Hawaii (HST)"
            UK -> "UK (GMT/BST)"
            CENTRAL_EUROPE -> "Central Europe (CET)"
            JAPAN -> "Japan (JST)"
            SYDNEY -> "Sydney (AEST)"
        }

    companion object {
        fun fromValue(value: String): AppTimezone? =
            entries.find { it.value == value }
    }
}
